solr/solrj/src/java/org/apache/solr/client/solrj/impl/LBHttpSolrClient.java

```diff
@@ -350,82 +351,69 @@ protected HttpSolrClient makeSolrClient(String server) {
    *
    * @throws IOException If there is a low-level I/O error.
    */
-  public Rsp request(Req req) throws SolrServerException, IOException {
-    Rsp rsp = new Rsp();
-    Exception ex = null;
+  public ResponseAndException request(Req req) throws SolrServerException, IOException {
     boolean isNonRetryable = req.request instanceof IsUpdateRequest || ADMIN_PATHS.contains(req.request.getPath());
-    List<ServerWrapper> skipped = null;
+    final int numDeadServersToTry = req.getNumDeadServersToTry();
+    List<ServerWrapper> skipped = new ArrayList<>(numDeadServersToTry);
 
     boolean timeAllowedExceeded = false;
     long timeAllowedNano = getTimeAllowedInNanos(req.getRequest());
     long timeOutTime = System.nanoTime() + timeAllowedNano;
-    for (String serverStr : req.getServers()) {
-      if (timeAllowedExceeded = isTimeExceeded(timeAllowedNano, timeOutTime)) {
-        break;
-      }
-      
+
+    //query in parallel
+    Optional<ResponseAndException> aliveResponseAndException = req.getServers().parallelStream().map(serverStr ->{
       serverStr = normalize(serverStr);
+
       // if the server is currently a zombie, just skip to the next one
       ServerWrapper wrapper = zombieServers.get(serverStr);
       if (wrapper != null) {
-        // System.out.println("ZOMBIE SERVER QUERIED: " + serverStr);
-        final int numDeadServersToTry = req.getNumDeadServersToTry();
-        if (numDeadServersToTry > 0) {
-          if (skipped == null) {
-            skipped = new ArrayList<>(numDeadServersToTry);
-            skipped.add(wrapper);
-          }
-          else if (skipped.size() < numDeadServersToTry) {
-            skipped.add(wrapper);
-          }
+        //only add to retry list if the number of dead servers to retry list hasn't been filled yet
+        if (skipped.size() < numDeadServersToTry) {
+          skipped.add(wrapper);
         }
-        continue;
+        return null;
       }
-      try {
-        MDC.put("LBHttpSolrClient.url", serverStr);
-        HttpSolrClient client = makeSolrClient(serverStr);
-
-        ex = doRequest(client, req, rsp, isNonRetryable, false, null);
-        if (ex == null) {
-          return rsp; // SUCCESS
-        }
-      } finally {
-        MDC.remove("LBHttpSolrClient.url");
+      else{
+        return getResponseAndException(req, isNonRetryable, true, serverStr, makeSolrClient(serverStr));
       }
-    }
+    }).findAny();
```

Comment:
We probably want to filter out Responses with exceptions since there might be a valid response that doesn't contain an exception.